### ISSUE TITLE
Add support for custom tag prefix for monorepo module

### ISSUE THREAD

**Issue by sumanmaity112** [serpro69/semver.kt#23]
I am new to this project. I was using https://github.com/alcideio/gradle-semantic-build-versioning/tree/master but doesn't works for monorepo.

Consider this project's structure:
```.
├── build.gradle.kts
├── gradle.properties
├── settings.gradle.kts
├── release                            <--- sub-module
│   ├── build
│   ├── build.gradle.kts
│   └── src
├── semantic-versioning          <--- sub-module
│   ├── build
│   ├── build.gradle.kts
│   └── src
└── spec                               <--- sub-module
    ├── build
    ├── build.gradle.kts
    └── src
    
```

All three sub-modules publish their own artifacts and maintains version independently. Now, I want to use tag something like: **release-v1.0.0**, **semantic-versioning-v0.2.9** etc.

Is there a possibility to support similar tagging process?

**Comment by serpro69** [comment]
Hi, @sumanmaity112 ,

Thank you for opening this issue. The project you're referring to seems like a "fork of a fork of a fork" of the project that I used some time ago as well, which later became unmaintained and unusable for me, hence this one was born :grinning:

Just to clarify, you want a different prefix for different tags, each tag being "linked" to each of the sub-modules basically?

If so, I think this would actually be a nice addition on top of https://github.com/serpro69/semver.kt/issues/17 , and it could also be a potential solution for some of the downsides I mentioned in that issue when it comes to versioning modules separately. 

This is currently not supported - you can customize the tag prefix for the entire git project, but not for individual sub-modules. But I'm very open to adding this.
If you'd like to work on a PR - please let me know. But I completely understand if you don't want to :) while the code here isn't complicated, and I try to have kdocs everywhere at least, there's still lack of proper "contributing documentation" that could be useful for someone new to the code. I'll try to find time pick this up myself otherwise.

**Comment by sumanmaity112** [comment]
Hi @serpro69, I'll look into the code first and then let you know if I'll able to work on it or not.

In very high level I am thinking config something like

```json
{
  "git": {
    "repo": {
      "directory": ".",
      "remoteName": "origin",
      "cleanRule": "tracked"
    },
    "message": {
      "major": "[major]",
      "minor": "[minor]",
      "patch": "[patch]",
      "ignoreCase": "true"
    }
  },
  "version": {
    "initialVersion": "0.1.0",
    "placeholderVersion": "0.0.0",
    "defaultIncrement": "patch"
  },
  "monorepo": [
    {
      "name": "abc",
      "tagPrefix": "abc-v"
    },
    {
      "name": "xyz",
      "tagPrefix": "xyz-v"
    }
  ]
}

```
Here _tagPrefix_ is an optional field, if not available then it'll use the config **from root.git.tag**.

Please add your opinion on this suggestion

**Comment by serpro69** [comment]
I've already started a bit ;) (see branch `multi_tagging` if you're interested)

That's similar to what I was thinking of as well and added new configuration param in the branch to get a feeling of how it would look like.
Though atm I think I'd rather reuse the whole `git.tag` configuration interface rather than adding a single standalone `tagPrefix` field to module configuration. So it would something like:

```json
{
  "git": {
    "repo": {
      "directory": ".",
      "remoteName": "origin",
      "cleanRule": "tracked"
    },
    "message": {
      "major": "[major]",
      "minor": "[minor]",
      "patch": "[patch]",
      "ignoreCase": "true"
    }
  },
  "version": {
    "initialVersion": "0.1.0",
    "placeholderVersion": "0.0.0",
    "defaultIncrement": "patch"
  },
  "monorepo": [
    {
      "name": "abc"
    },
    {
      "name": "xyz",
      "tag": {
         "tagPrefix": ...
         "separator": ...
      }
    }
  ]
}

```

The `.monorepo[].tag` configuration will be optional. If it's set for a submodule, then it will be used, otherwise the `.git.tag` configuration is used.
And `.git.tag` config is always used for the root project.

**Comment by serpro69** [comment]
Will try to prioritize this in free time, need it also for [kotlin-faker](https://github.com/serpro69/kotlin-faker) :grinning:

**Comment by serpro69** [comment]
Got a bit busy with work past few weeks, but will try to prioritize this in free time - also need this for [kotlin-faker](https://github.com/serpro69/kotlin-faker) 😀

**Comment by serpro69** [comment]
After a bit of refactoring of configs, tested it manually and looks like we're getting there :grin: 

```bash
tmp.FyBioe1JSW on  master …
➜ git lg
* c996d58 - (44 minutes ago) Configure modules - Serhii Prodan (HEAD -> master)
* 66e5410 - (47 minutes ago) Fix compilatio - Serhii Prodan (tag: v0.2.0)
* 0d29a8a - (48 minutes ago) Add foo and bar submodules - Serhii Prodan
* 64bad4c - (50 minutes ago) Add semver plugin - Serhii Prodan (tag: v0.1.0)
* 5fb1561 - (51 minutes ago) initial commit - Serhii Prodan

tmp.FyBioe1JSW on  master via 🅺 v1.9.22 …
➜ ./gradlew clean tag -Prelease
Configure module :foo
Modules:
 { "path": ":foo", "sources": "src", "tag": { "prefix": "foo-v", "separator": "", "useBranches": "false" } }
Configure module :bar
Modules:
 { "path": ":foo", "sources": "src", "tag": { "prefix": "foo-v", "separator": "", "useBranches": "false" } }
 { "path": ":bar", "sources": "src", "tag": { "prefix": "bar-v", "separator": "", "useBranches": "false" } }

> Task :tag
Calculated next version: 0.3.0
Next version: 0.3.0
Prefix: v

> Task :bar:tag
Calculated next version: 0.1.0
Next version: 0.1.0
Prefix: bar-v

> Task :foo:tag
Calculated next version: 0.1.0
Next version: 0.1.0
Prefix: foo-v

> Task :lib:tag
Calculated next version: 0.3.0
Next version: 0.3.0
Prefix: v
Tag v0.3.0 already exists in project

BUILD SUCCESSFUL in 943ms
7 actionable tasks: 4 executed, 3 up-to-date

tmp.FyBioe1JSW on  master via 🅺 v1.9.22 …
➜ git lg
* c996d58 - (43 minutes ago) Configure modules - Serhii Prodan (HEAD -> master, tag: v0.3.0, tag: foo-v0.1.0, tag: bar-v0.1.0)
* 66e5410 - (46 minutes ago) Fix compilatio - Serhii Prodan (tag: v0.2.0)
* 0d29a8a - (47 minutes ago) Add foo and bar submodules - Serhii Prodan
* 64bad4c - (48 minutes ago) Add semver plugin - Serhii Prodan (tag: v0.1.0)
* 5fb1561 - (50 minutes ago) initial commit - Serhii Prodan

```

Now need to add proper tests to verify various edge-cases, etc. Hopefully won't take long :crossed_fingers:

**Comment by serpro69** [comment]
This is available in [v0.10.0](https://github.com/serpro69/semver.kt/releases/tag/v0.10.0) 
Please try it out and feel free to open another issue if you discover any bugs :+1:
